### ISSUE TITLE
Coq repository contains too many files at root

### ISSUE THREAD

**Issue by Zimmi48** [coq/coq#6137]
#### Context

Yesterday, @letouzey told me privately that he was against my `default.nix` PR #6123 or at least against this file being added at the root of the repository (which is a pity since it will work best there). He was ready to make the same argument against an OPAM file at the root of the repository (cf issue #5758).

The overall argument is that the Coq repository is for Coq sources (strictly speaking) and there is already too much clutter at the root of the repository. I disagree with the first part but I fully agree with the second part. This issue is thus trying to propose to reduce the number of files that are at the root of the repository.

#### Motivation

My personal motivation for reducing the number of files at the root of the repository is that people (especially newcomers) find more easily the important information (such as the README).

#### Steps

I am going to submit a first PR removing two useless files and moving what can be into a `.github/` sub-directory. But I think we should also consider:

- [ ] Moving all `Makefile` but the main one to a sub-directory (and `configure.ml` and `install.sh` as well).
- [ ] Moving `META.coq` to a sub-directory if that doesn't prevent installing it properly.
- [ ] Moving `.bintray.json` in `dev` if possible.
- [ ] Archiving the `COMPATIBILITY` file as it seems it isn't used so much anymore and the `README` says that it is the `CHANGES` file that document the incompatibilities (to be discussed).
- [ ] Moving all source directories (`API` `engine` `grammar` `interp` `intf` `kernel` `lib` `library` `parsing` `pretyping` `printing` `proofs` `stm` `tactics` `toplevel` and `vernac`) to a `src` sub-directory. (Remark I didn't list here `checker` `config` `dev` `doc` `ide` `plugins` `test-suite` `theories` and `tools` which have special status.)

**Comment by maximedenes** [comment]
> the Coq repository is for Coq sources (strictly speaking)

I also disagree with that. I don't see why it would not contain material that is helpful for developers. If we follow that path, we should also remove all developer documentation and testing infrastructure, as they are also not strictly speaking part of the source code.

> here is already too much clutter at the root of the repository

I have nothing against a reorganization, but I don't find the situation particularly bad. It is fairly easy to see what each root file or folder is for.

Don't forget to test all build scripts if you move stuff, but IMHO our time would be better used cleaning the actual code. Seriously, look at the plugins folder. The priority is probably not a cleanup of the layout of the file tree, but rather of their *content*...

**Comment by SkySkimmer** [comment]
> Moving all Makefile but the main one to a sub-directory (and configure.ml and install.sh as well).

This could in principle be put in config/ as a directory for directing the Coq build.

**Comment by Zimmi48** [comment]
> It is fairly easy to see what each root file or folder is for.

I think you are blinded by your long experience with the Coq sources. It wasn't clear to me, at least at the very beginning, what each folder was and which contained the Coq source code and which didn't.

> Don't forget to test all build scripts if you move stuff, but IMHO our time would be better used cleaning the actual code. Seriously, look at the plugins folder. The priority is probably not a cleanup of the layout of the file tree, but rather of their content...

I'm not going to do it personally because I am way to uncertain when it comes to build systems. That's why I've opened this issue instead. As for what the priority is, I kind of not like this type of counter-argument because obviously you are comparing a minor layout question with some major clean-up which will take years and years.

> > Moving all Makefile but the main one to a sub-directory (and configure.ml and install.sh as well).
>
> This could in principle be put in config/ as a directory for directing the Coq build.

Indeed, that's what I was thinking as well. But I won't do it myself because I would be pretty sure to break stuff. If someone wants to give it a try, please do.

**Comment by ejgallego** [comment]
I see little gain from most of this patch and a big breakage for many existing setups and patches, even if I agree that the tree is a bit hard to navigate a first.

**Comment by ejgallego** [comment]
I disagree with @letouzey about the nix file. It is useful and doesn't harm anybody.

Some things we could improve indeed:
- move the makefiles.
- `install.sh`
- `man` to `doc/man`
- build in `_build`, remove `bin`
- 3x `INSTALL` to one `INSTALL`
- `LICENSE` / `COPYRIGHT` / `CREDITS`.
- autogenerate `.merlin` ?

**Comment by ejgallego** [comment]
Also we could put in `src` the sources, but maybe of the part that is "modular". In that case we'd have from `kernel` up to `stm`. `toplevel` for example is standalone.

**Comment by Zimmi48** [comment]
Indeed, that's why I had not listed `ide` in the list of directories that could be moved to sources. I hadn't realized that the same reasoning applied to `toplevel`.

**Comment by ejgallego** [comment]
It is just a non-working "brouillon", but this how I envision the tree in the very future Dune branch: https://github.com/ejgallego/coq/tree/try_src

[Why Dune: because it allows me to place components more freely without having to edit makefiles and paths which I always get wrong]

**Comment by Zimmi48** [comment]
It would look perfect like this (and I strongly support the move to Dune with the flexibility it brings).
Why do you need to have `bin/` checked in the repository though?

**Comment by ejgallego** [comment]
So the idea is that `core` contains the "library" parts of coq, and `bin` the sources for the binaries (workers, toplevel, etc..)

How should we name `bin` then?

**Comment by ejgallego** [comment]
So yeah compiling `core` produces all the `coq.kernel` `coq.engine` etc... components, but no binary.

**Comment by Zimmi48** [comment]
I don't mind `bin/` but it was empty in your branch, so I was not sure what its purpose was.

**Comment by ejgallego** [comment]
The nice thing about Dune is that you specify "my object needs `coq.proofs`" and Dune will use `core/proofs` if it is in the source tree, or else use `ocamlfind` to locate it. This is _extremely_ convenient for a project such as coq.

**Comment by ejgallego** [comment]
> I don't mind bin/ but it was empty in your branch, so I was not sure what its purpose was.

Yeah it was a sketch sorry, the purpose is to have the sources for the binaries themselves, but indeed maybe the name is confusing.

**Comment by ejgallego** [comment]
After #13617 this can be solved by simply moving directories around, I propose we implement a plan or close this issue.

**Comment by Zimmi48** [comment]
I propose the following plan:
- `core/` contains all the sources of `coq-core` (thus including binaries like `coqtop` but also `coqpp` / `coqchk` and `plugins/` as long as they are still part of the `coq-core` package);
- thus apart from `core/` the only directories we would keep at the root are `.github/`, `config/`, `dev/`, `doc/`, `ide/`, `man/` (although I would favor moving it under `doc/`), `test-suite/`, `theories/` and `user-contrib/`.

**Comment by ejgallego** [comment]
I like this plan, tho I'd be a tad more conservative for now and keep `coqpp/coqchk/plugins` in the root folder for now, as even if they are part of the coq-core package, they are quite decoupled from the "core" libs.

Moving everything in coq-core to core would require also moving configure, dev, parts of man, and `config`, so that'd be possible too if we'd like to do a true split.
